staging/kos/pkg/controllers/connectionagent/connectionagent.go

```diff
@@ -0,0 +1,1091 @@
+		resyncPeriod,
+		fromFieldsSelectorToTweakListOptionsFunc(localAttWithAnIPSelector))
+
+	ca.localAttsInformer.AddIndexers(map[string]k8scache.IndexFunc{attMACIndexName: attachmentMACAddr})
+
+	ca.localAttsInformer.AddEventHandler(k8scache.ResourceEventHandlerFuncs{
+		AddFunc:    ca.onLocalAttAdded,
+		UpdateFunc: ca.onLocalAttUpdated,
+		DeleteFunc: ca.onLocalAttRemoved,
+	})
+}
+
+func (ca *ConnectionAgent) onLocalAttAdded(obj interface{}) {
+	att := obj.(*netv1a1.NetworkAttachment)
+	glog.V(5).Infof("Local NetworkAttachments cache: notified of addition of %#+v", att)
+	ca.queue.Add(kosctlrutils.AttNSN(att))
+}
+
+func (ca *ConnectionAgent) onLocalAttUpdated(oldObj, newObj interface{}) {
+	oldAtt := oldObj.(*netv1a1.NetworkAttachment)
+	newAtt := newObj.(*netv1a1.NetworkAttachment)
+	glog.V(5).Infof("Local NetworkAttachments cache: notified of update from %#+v to %#+v", oldAtt, newAtt)
+	ca.queue.Add(kosctlrutils.AttNSN(newAtt))
+}
+
+func (ca *ConnectionAgent) onLocalAttRemoved(obj interface{}) {
+	peeledObj := kosctlrutils.Peel(obj)
+	att := peeledObj.(*netv1a1.NetworkAttachment)
+	glog.V(5).Infof("Local NetworkAttachments cache: notified of removal of %#+v", att)
+	ca.queue.Add(kosctlrutils.AttNSN(att))
+}
+
+func (ca *ConnectionAgent) waitForLocalAttsCacheSync(stopCh <-chan struct{}) (err error) {
+	if !k8scache.WaitForCacheSync(stopCh, ca.localAttsInformer.HasSynced) {
+		err = fmt.Errorf("Caches failed to sync")
+	}
+	return
+}
+
+func (ca *ConnectionAgent) handlePreExistingIfcs() error {
+	err := ca.handlePreExistingLocalIfcs()
+	if err != nil {
+		return err
+	}
+
+	err = ca.handlePreExistingRemoteIfcs()
+	return err
+}
+
+func (ca *ConnectionAgent) handlePreExistingLocalIfcs() error {
+	allPreExistingLocalIfcs, err := ca.netFabric.ListLocalIfcs()
+	if err != nil {
+		return fmt.Errorf("Failed initial local network interfaces list: %s", err.Error())
+	}
+
+	for _, aPreExistingLocalIfc := range allPreExistingLocalIfcs {
+		ifcMAC := aPreExistingLocalIfc.GuestMAC.String()
+		ifcOwnerAtts, err := ca.localAttsInformer.GetIndexer().ByIndex(attMACIndexName, ifcMAC)
+		if err != nil {
+			return fmt.Errorf("Indexing local network interface with MAC %s failed: %s", ifcMAC, err.Error())
+		}
+
+		if len(ifcOwnerAtts) == 1 {
+			// If we're here there's a local attachment which should own the interface because their
+			// MAC addresses match. Hence we add the interface to the attachment state
+			ifcOwnerAtt := ifcOwnerAtts[0].(*netv1a1.NetworkAttachment)
+			attVNI, attNSN := ifcOwnerAtt.Spec.VNI, kosctlrutils.AttNSN(ifcOwnerAtt)
+			attState := ca.getAttState(attNSN)
+			if attState == nil {
+				attState := &attachmentState{
+					vnStateVNI: attVNI,
+				}
+				ca.setAttState(attNSN, attState)
+			}
+			_, firstLocalAttInVN := ca.updateVNState(attState, attVNI, attNSN, ifcOwnerAtt.Spec.Node)
+			if firstLocalAttInVN {
+				glog.V(2).Infof("VN with ID %d became relevant: an Informer has been started", attVNI)
```

Comment:
Why log here rather than closer to where the fact is determined?

Reply:
The actual fact takes place in a section protected by a mutex. I'd rather do only what is strictly necessary when holding a Lock. I can move the logging one method down because that's still outside of the critical section.

Reply:
that's what I meant by "closer"

Reply:
Fixed in 08d7d06.